Remove need for expression languages (use JsonPath only) (#97)
serverlessworkflow/specification: specification.md

## Specification Goals

At the core of the Serverless Workflow specification is described via its [JSON Schema](schema/workflow.json).
This schema defines the workflow model. It can also be used for generation of many different artifacts

> such as APIs and SPIs and SDKs. Currently the specification provides SDKs for Go and Java and we hope to 

Reviewer: ```suggestion
such as APIs and SPIs and SDKs. Currently the specification provides SDKs for [Go](https://github.com/serverlessworkflow/sdk-go) and [Java](https://github.com/serverlessworkflow/sdk-java) and we hope to 
```

Author: fixed

expand this list in the future. 

> We also strives to provide a Technology Compatibility Kit (TCK) which provides a suite of tests usec for

Reviewer: ```suggestion
We also strive to provide a Technology Compatibility Kit (TCK) which provides a suite of tests used for
```

Author: fixed

The defined workflow model is a declarative language that can be used to model small or complex orchestrations
for event-driven, serverless applications.

### Workflow Data

Serverless Workflow data is represented in [JSON](https://www.json.org/json-en.html) format.
Data flow during workflow execution can be divided into:

- [Workfow data input](#Workflow-data-input)
- [Event data](#Event-data)
- [Action data](#Action-data)
- [Information passing between states](#Information-passing-Between-States)
- [State information filtering](#State-information-filtering)
- [Workflow data output](#Workflow-data-output)

### Workflow Data Expressions

Workflow model parameters may use expressions to access the JSON data. 
Note that different data filters play a big role as to which parts of the state data is queried against. Reference the 
[State Information Filtering](#State-Information-Filtering) section for more information.

All expressions must follow the [JsonPath](https://github.com/json-path/JsonPath) format and can be evaluated with a JsonPath expression evaluator. 

> JSONPath expressions can traverse the workflow JSON data. 

Reviewer: ```suggestion
JsonPath expressions can traverse the workflow JSON data. 
```

Author: fixed and also updated all places to change from "JSONPath" to "JsonPath"

All expressions are written inside double braces:

```text
{{ expression }}
```

Expressions should be resolved and the result returned exactly where the expression is written.

To show some expression examples, let's say that we have the following JSON data:

```json
{
  "firstName": "John",
  "lastName" : "Doe",
  "age"      : 26,
  "address"  : {
    "streetAddress": "Naist street",
    "city"         : "Nara",
    "postalCode"   : "630-0192"
  },
  "phoneNumbers": [
    {
      "type"  : "iPhone",
      "number": "0123-4567-8888"
    },
    {
      "type"  : "home",
      "number": "0123-4567-8910"
    }
  ]
}
```

We can use an expression inside a string type parameter, for example:

```json
{
    "paramName": "Hello {{ $.firstName }} {{ $.lastName }}"
}
```

> would set the value of "paramName" to "Hello John Doe".

Reviewer: ```suggestion
would set the value of `paramName` to `Hello John Doe`.
```

Author: fixed

Expressions can also be used to select a portion of the JSON data, this is in particularly useful for data filters. For example:

```json
{
   "stateDataFilter": {
       "dataOutputPath": "{{ $.phoneNumbers }}"
   }
}
```

Would set the data output of the particular state to:

```json
[
   {
      "type" : "iPhone",
      "number" : "0123-4567-8888"
   },
   {
      "type" : "home",
      "number" : "0123-4567-8910"
   }
]
```

Some parameters require an expression to resolve to a boolean (true / false). In this case JsonPath expressions can also be used.
The expression should evaluate to true, if the result contains a subset of the JSON data, and false if it is empty. For example:

```json
{
   "isAdult" : "{{ $.[?(@.age  > 18)] }}"
}
```

> would set the value if "isAdult" to true, as the expression returns a non-empty subset of the JSON data. Similarly 

Reviewer: ```suggestion
would set the value if `isAdult` to true, as the expression returns a non-empty subset of the JSON data. Similarly 
```

Author: fixed

```json
{
   "isChild" : "{{ $.[?(@.age  < 18)] }}"
}
```

> would return false as the result of the JSONPath expression is empty.

Reviewer: ```suggestion
would return false as the result of the JsonPath expression is empty.
```

Author: fixed

JsonPath also includes a limited set of built-in functions that can be used inside expressions. For example the expression

```json
{
   "phoneNums" : "{{ $.phoneNumbers.length() }}"
}
```

> would set the value of 'phoneNums' to "2".

Reviewer: ```suggestion
would set the value of `phoneNums` to `2`.
```

Author: fixed

As previously mentioned, expressions are evaluated against the subset of data that can see. For example 

> the "parameters" param of the [functionRef definition](#FunctionRef-Definition) can evaluate expressions 

Reviewer: ```suggestion
the `parameters` param of the [functionRef definition](#FunctionRef-Definition) can evaluate expressions 
```

Author: fixed

only against the data that is available to the [action](#Action-Definition) it belongs to.
One thing to note here are the top-level [workflow definition](#Workflow-Definition) parameters. Expressions defined
in them can only be evaluated against the initial [workflow data input](#Workflow-Data-Input) JSON.
For example let's say that we have a workflow data input of:

```json
{
   "inputVersion" : "1.0.0"
}
```

we can then define a workflow definition:

```json
{
   "id": "MySampleWorkflow",
   "name": "Sample Workflow",
   "version": "{{ $.inputVersion }}"
}
```

> which would set the workflow version to "1.0.0".

Reviewer: ```suggestion
which would set the workflow version to `1.0.0`.
```

Author: fixed
